### DISCUSSION TITLE
channel HDMI output to blackhole on mac

### DISCUSSION THREAD

**Discussion by theshuklz** [ExistentialAudio/BlackHole#40] [Feature Requests]
My speakers are connected to the monitor. when I connect my mac to the monitor, normally it doesn't allow adjusting volume mac. Soundflower had this feature where I could provide the desktop output to it and then I can control it.
Is it possible to do the same in Blackhole, if yes, how?

**Comment by devinroth** [comment]
Not currently. Another piece of the puzzle is needed to map BlackHole's output to a different output.

**Comment by devinroth** [comment]
But this is an often requested feature. I'll keep this up.

**Comment by kimsyversen** [comment]
This would be really great feature!

+1

**Comment by ByteSkater** [comment]
+1, as currently the only solution I found for this is a commercial product and soundflower also doesn't work anymore.

**Comment by emmaworley** [comment]
This would be great.

**Comment by gosuto-inzasheru** [comment]
+1

**Comment by aryadaroui** [comment]
This would be great. +1

**Comment by no-Legend** [comment]
This would be great for me too. +1

**Comment by egyedia** [comment]
+1

**Comment by redstrike** [comment]
+1
I plugged in USB Audio Device into my MacBook, and cannot find any way to control its volume.

**Comment by jbuecker** [comment]
+1

**Comment by alexanderchuranov** [comment]
+1
I have a similar scenario: purchased a professional audio interface and suddenly the usual mac up/down volume buttons do not work (device offers its own volume controls). I'd like to route all applications to Blackhole and from there to the audio interface. Loopback certainly offers that option, but it is not open-source.

**Comment by igem0n** [comment]
+1

**Comment by WickedColdfront** [comment]
This would be great!

**Comment by MichaelGrundie** [comment]
+10 I want :)

**Reply by markaudacity** [reply]
+1, number one feature I'm looking for: audio out my MacBook Pro's speakers with volume control, simultaneously branched to another app.

**Comment by bockbilbo** [comment]
+1

I use the speaker of my external monitor and this would allow using the volume controls of mac to change the volume :)

Thank you!

**Comment by inter-faced** [comment]
+1 adjusting volume control with media keys would be a major feature for me. Too bad apple make this so cumbersome

**Comment by carlosfilhoadm** [comment]
+1

Yes, please. This feature would prove itself very useful for a lot of users.

Thank you.

**Comment by PRosenb** [comment]
As a workaround, audio output of BlackHole can be redirected to the output device with a simple Python script.

[forwardAudio.py](https://gist.github.com/PRosenb/2ce5159a334af203fca7ef06501775f0)

I hope it's useful.

**Comment by rvlak** [comment]
+1

**Comment by baptisteArno** [comment]
+1

**Comment by Coriou** [comment]
+1 =)

**Comment by gosuto-inzasheru** [comment]
Isn't it weird Apple hasn't picked this up? Should we contact them somehow?

Judging by the popularity of this everybody just wants to control their peripherals' volume using their keyboard, which is more than logical :-)

**Comment by fchu** [comment]
+1

**Comment by jcperez-iti-upv** [comment]
+1

**Reply by yingwuzhao** [reply]
+1

**Comment by briankendall** [comment]
I hope this sort of self-advertising isn't considered inappropriate, but I wrote a virtual audio device a little while ago for precisely this purpose uncreatively called [Proxy Audio Device](https://github.com/briankendall/proxy-audio-device). I too had an audio interface whose manufacturer couldn't be bothered to allow the system volume to work, and I wasn't having it.

I'll leave that link with the caveat that it's the first audio driver I've ever written, and while I haven't abandoned it I don't have the time to support that I might otherwise like. So if another dev takes up the mantle of providing a driver that can do this, I'd be quite happy! I want people to have the best experience possible, and I worry that I'm not the best person to provide that.

(Also, my driver does some stuff that's _probably_ pretty hacky. If someone who knows what they're doing a bit better than me took a stab at it, that might be for the best!)

**Reply by rodrigolopezguerra** [reply]
Great ! This actually worked like a charm !!

**Comment by devinroth** [comment]
Thanks for sharing Brian. I’m sure it will be helpful for some.

**Comment by karthicraghupathi** [comment]
Great work. Thank you!

+1 for this feature.

**Comment by saucebing** [comment]
+1 Now when I want to adjust the volume I need to switch to the internal speakers and then switch back to multi-output device. The volume adjustment is necessary.

**Comment by bwagner** [comment]
+1: I'd need it with as little latency as possible (using it for looping live).

**Comment by jcrystal** [comment]
+1. The ability to monitor specific BlackHole channels via physical hardware would be so great.

**Comment by paulreimer** [comment]
If I understand correctly, part of the solution proposed for this issue would also make it possible to monitor the output without needing to create a Multi-Output Device? (that's the feature I'm looking for).

Sometimes I want to monitor but mostly I don't, and (currently) I need to switch audio devices to do that (causing hiccups and requiring selecting the new device in all apps that are using it).

**Comment by alexthedamager** [comment]
+1 this would be really awesome, especially if it could be configured using a menu bar app.

**Comment by itsjoshpark** [comment]
+1 Having this exact scenario. This feature would be immensely helpful.

**Comment by n8gray** [comment]
Same scenario: monitor without volume controls, but I also find lots of value in random uses for tools like this.  I'm currently using **Hosting AU** to forward from BH to the monitor.  It works, but it doesn't handle changes like disconnections and sleep well.  I've been playing with Apple's **CAPlaythrough** sample code and it works, but there's a sync problem and after about 40 minutes the ring buffer starts to clip.  I'd like to learn how to fix it, but it's hard to find the time. 

EDIT: I tried @briankendall's [Proxy Audio Device](https://github.com/briankendall/proxy-audio-device) previously [mentioned](https://github.com/ExistentialAudio/BlackHole/issues/40#issuecomment-642743519) and it works like a charm! If this is your use case I'd highly recommend it.

**Reply by niugm** [reply]
I am also trying to use CAPlaythrough, but I met the same problem. Have you solved it?

**Reply by n8gray** [reply]
No, once I found Proxy Audio Device my problem was solved so I stopped working on it.

**Comment by tspangler** [comment]
+1, but in the meantime, @briankendall's [Proxy Audio Device](https://github.com/ExistentialAudio/BlackHole/issues/40#issuecomment-642743519) driver works beautifully, and none of the Soundflower lag I grew accustomed to!

**Comment by shonenxnaifu** [comment]
+1 this would be nice feature!

**Comment by abishekmuthian** [comment]
> +1, but in the meantime, @briankendall's [Proxy Audio Device](https://github.com/ExistentialAudio/BlackHole/issues/40#issuecomment-642743519) driver works beautifully, and none of the Soundflower lag I grew accustomed to!

Audio seems to be lost after sleep, how are you handling it with Proxy Audio Device?

**Reply by uloco** [reply]
came here because of this issue i'm having all the time :/

**Comment by gnassel** [comment]
+1

**Comment by Maximilian-Pichler** [comment]
thanks, @devinroth for this awesome piece of software!
this feature would make me instantly switch from soundflower

**Comment by vsguts** [comment]
+1 really need this feature!

**Comment by aporzio1** [comment]
This would be awesome

**Comment by jacobsch** [comment]
+1, I agree with @theshuklz and would definitely switch over to BlackHole if this feature was supported. Right now I am using eqMac to control the output volume on some outputs that mac won't let you change (hdmi & usb audio sources for me). I would highly advise looking at [SoundflowerBed](https://github.com/mLupine/SoundflowerBed) for inspiration on how to implement a minuscule UI to allow these types of configurations once BlackHole supports this ability.

**Comment by pdefoy** [comment]
This is a must feature.
In the mean time, I am using Sound Source, it is not free but works great.
[SoundSource](https://rogueamoeba.com/soundsource/)

**Comment by JokerDen** [comment]
Yes, please! I know the app eqMac that allows such functionality (change overall volume for HDMI audio output) put it has a lot of other unnecessary functionality unlike Blackhole.

**Comment by saulpower** [comment]
+1000

**Comment by DamianoP** [comment]
+1

**Comment by romansavrulin** [comment]
Theres's also [ladiocast](https://apps.apple.com/ru/app/ladiocast/id411213048?mt=12) app to proxy audio from several devices to another with basic mixing options

**Comment by Marcello1099** [comment]
+1

**Comment by senovsky** [comment]
+1

**Comment by bockbilbo** [comment]
I can't believe its been over 3 years since this request. I was able to resolve it Ventura using this - https://github.com/briankendall/proxy-audio-device

**Comment by devinroth** [comment]
Wow. What a rude thing to say. You’re welcome for developing free open source software.

Devin

**Comment by bockbilbo** [comment]
Devin, you misunderstood my comment. I am and will always be grateful of
open source developers and understand the nature and lifecycle of open
source projects. My comment around the two years was in terms of me being
surprised of the time it has passed since I discovered the project.

**Comment by devinroth** [comment]
Sure doesn't come across that way. Sounds more like disappointment in the fact that I haven't addressed this issue.

**Comment by devinroth** [comment]
And to offer an explanation of why I haven't addressed this issue. Along with juggling family, work, and open source, I've come to the conclusion that this feature should not be part of BlackHole. BlackHole is an audio driver. Adding this ability would be beyond the scope of a driver in my opinion. I can discuss that in further detail if you'd like. I've created solutions to this in various projects I've worked on but in all instances the part of the code that handles outputting audio to a different audio device is done outside the driver. I'm considering building an releasing an app that will specifically solve this problem but haven't got around to it yet.

**Reply by bockbilbo** [reply]
Thanks for sharing your thoughts. I don't think anyone is disappointed by the fact that the issue has not been addressed. If anything, we are grateful for your work!

When I posted my original +1 to this feature request in 2020 I already guessed that the feature was unlikely to be prioritized as it falls outside the scope of project. Back then I was able to redirect audio sources to different audio devices using SoundFlower and today I did it with proxy-audio-device.

In my honest opinion, you should close this request. :)
